_posts/2022-07-24-opensearch-plugin-zips-now-in-maven-repo.markdown

```diff
@@ -0,0 +1,74 @@
+---
+layout: post
+title:  "OpenSearch plugin zips now in Maven repo"
+authors: 
+  - prudhvigodithi
+date: 2022-07-24 01:01:01 -0700
+categories: 
+  - technical-post
+twittercard:
+  description: "Details on how to consume OpenSearch plugin zips from maven repo and the process involved in shipping them to the maven repo."
+---
+
+### Motivation and Current Challenges:
+
+The plugins zips that are used as dependencies cannot be downloaded dynamically during runtime, as they are not part of version controlled maven system. Only the respective Java jars are available to download using maven coordinates. This forces a user to now to use a dependency plugin zip of some random version built on a developer desktop, which is not reliable, also the tests executed as part of this build zip cannot be accurate, so eventually for few plugins the local build zip is forced to check in to the plugin repo to facilitate the plugin availability for dependency, this is one of the challenge faced by the community. Coming to the download part of plugins as separate isolated components via a cached mechanism like maven is not possible, with this challenge users had to  pre-bake the random built zip to the project.
+
+Starting with release `2.1.0` OpenSearch plugin zips are now signed and published to central maven repo, with `groupID` as [org.opensearch.plugin](https://repo1.maven.org/maven2/org/opensearch/plugin/). These zips can now be fetched as individual components by directly downloading using maven coordinates or by clickstream from central maven repo which can be later cached to the desired user local maven repo. The snapshot version of the zips can be fetched from nexus maven repo with the same `groupID` as [org.opensearch.plugin](https://aws.oss.sonatype.org/content/repositories/snapshots/org/opensearch/plugin/).
+
+### Maven publish task with custom gradle plugin.
+
+The publishing of zips to central maven repo had been possible with custom gradle plugin `opensearch.pluginzip`.
+Starting with release `2.1.0` all OpenSearch gradle supported plugins will have custom gradle plugin `opensearch.pluginzip` applied to the gradle project, this will create a new task `publishPluginZipPublicationToZipStagingRepository` which does all the heavy lifting in identifying the distribution plugin zip, setting the maven coordinates, generating the POM file, updating the user passed POM fields and publishing it to maven repo. More details and inner workings of the plugin can found at [opensearch-plugins](https://github.com/opensearch-project/opensearch-plugins/blob/main/BUILDING.md#opensearchpluginzip).
```

Comment:
```suggestion
Using OpenSearch plugin zips through Maven offers the following benefits:

- Plugins zip in the central Maven repo is already signed with `.asc`, `.md5`, `.sha1`, `.sha256`, `.sha512` extensions.
- Users are no longer required to to check in zips to any `src/` files, since zips can be fetched with the right `groupID`, `artifactID` and `version`.
- Tests and CI workflows can directly run against the Maven repo, instead of requiring a manual download. 
```

Reply:
- Tests and CI workflows can directly run against the zips from Maven repo, instead of requiring a manual download.

Reply:
In the first list item, change to "Plugin zips in the central Maven repo are already signed with...". Add "and" after the last comma. In the second list item, remove the comma after "files" and add a comma after "'artifactID'". In the third list item, change to "Tests and CI workflows can run directly against the Maven repo instead of requiring a manual download."